File: VideoApp/Models/ViewModel/VideoViewModel.cs
using System.Collections.Generic;
using System.Linq;

namespace VideoApp.Models.ViewModel
{
    public class VideoViewModel
    {
        private List<Video> _videos;
        public List<Video> Videos
        {
            get { return _videos ?? (_videos = new List<Video>()); }
            set { _videos = value; }
        }

        private List<Author> _authors;
        public List<Author> Authors
        {
            get { return _authors ?? (_authors = new List<Author>()); }
            set { _authors = value; }
        }

        private List<Genre> _genres;
        public List<Genre> Genres
        {
            get { return _genres ?? (_genres = new List<Genre>()); }
            set { _genres = value; }
        }

        public void LoadData()
        {
            using (var videocontext = new VideoContext())
            {
                videocontext.Configuration.ProxyCreationEnabled = false;
                _videos = videocontext.Video.ToList();

                _authors = videocontext.Authors.ToList();
                _genres = videocontext.Genres.ToList();
            }
        }

        public void DeleteVideo(int id)
        {
            using (var videocontext = new VideoContext())
            {
                var entity = videocontext.Video.First(x => x.VideoID == id);
                videocontext.Video.Remove(entity);
                videocontext.SaveChanges();
            }
        }

        public void AddVideo(Video video)
        {
            using (var videocontext = new VideoContext())
            {
                videocontext.Configuration.ValidateOnSaveEnabled = false;
                videocontext.Video.Add(video);
                videocontext.SaveChanges();
            }
        }

        public void UpdateVideo(Video video)
        {
            using (var videocontext = new VideoContext())
            {

                videocontext.Configuration.ValidateOnSaveEnabled = false;
                foreach (var entity in videocontext.Video.Where(x => x.VideoID == video.VideoID))
                {
                    entity.DatePremiere = video.DatePremiere;
                    entity.GenreID = video.GenreID;
                    entity.Title = video.Title;
                    entity.GenreID = video.GenreID;
                }
                videocontext.SaveChanges();
            }
        }
    }
}

File: VideoApp/Scripts/ViewModels/VideoViewModel.js

/// <reference path="~/Scripts/System/knockout-2.2.0.js" />
/// <reference path="~/Scripts/System/knockout.mapping-latest.js" />
/// <reference path="~/Scripts/System/jquery-2.0.0.min.js" />

$(document).ready(function () {

});

function DatePicker() {
    $.datepicker.setDefaults($.datepicker.regional['ru']);
    $("input.date").datepicker();
}

function createNewRow() {
    $("#table").hide();
    $("#createTable").show();
}

function cancelNewRow() {
    $('#video tbody tr.NewRow:eq(0)').remove();
}

function videoRow(data, genres, authors) {
    this.VideoID = ko.observable(data.VideoID);
    this.AuthorID = ko.observable(data.AuthorID);
    this.Title = ko.observable(data.Title);
    this.DatePremiere = ko.observable(data.DatePremiere);
    this.GenreID = ko.observable(data.GenreID);
    this.genre = ko.observable(data.Genre.KindGenre);
    this.author = ko.computed(function () {
        return data.Author.FirstName + " " + data.Author.LastName;
    });
}

function VideoViewModel() {
    var self = this;
    //Data
    self.Videos = ko.observableArray([]);
    self.Genres = ko.observableArray([]);
    self.Authors = ko.observableArray([]);

    //Actions
    self.createRow = function () {
        if (!$('#video tbody tr').is(".NewRow")) {
            $('#video').append($('#createTable tbody tr.NewRow:eq(0)').clone());
            var cancelRow = document.getElementById('cancel');
            cancelRow.onclick = cancelNewRow;
            $('#DatePrimiere').datepicker();
            $('.add').click(self.addVideo);
        }
    };

    self.addVideo = function () {
        var JSON = ko.toJSON({
            Title: $('#Title').val(),
            DatePremiere: $("#DatePrimiere").val(),
            AuthorID: $('#AuthorID').val(),
            GenreID: $('#GenreID').val()
        });
        $.ajax('/Home/AddRow', {
            data: JSON,
            type: "POST",
            contentType: "application/json",
            success: function (data) {
                self.Videos.push(new videoRow(data, self.Genres, self.Authors));
                cancelNewRow();
            }
        });
    };

    self.editRow = function (item) {
        var index = self.Videos.indexOf(item);
        $('#video tbody tr.text:eq(' + (index) + ')').hide();
        $('#video tbody tr.edit:eq(' + (index) + ')').show();
    };

    self.updateRow = function (item) {
        var index = self.Videos.indexOf(item);
        var JSON = ko.toJSON({
            AuthorID: item.AuthorID().AuthorID(),
            GenreID: item.GenreID().GenreID(),
            DatePremiere: item.DatePremiere,
            Title: item.Title,
            VideoID: item.VideoID
        });
        $.ajax('/Home/UpdateRow', {
            data: JSON,
            type: "POST",
            contentType: "application/json",
            success: function (data) {
                var mappedVideo = $.map(data, function (item) { return new videoRow(item, self.Genres, self.Authors) });
                self.Videos(mappedVideo);
                $('#video tbody tr.text:eq(' + (index) + ')').show();
                $('#video tbody tr.edit:eq(' + (index) + ')').hide();
            }
        });
    };

    self.removeRow = function (item) {
        var JSON = ko.toJSON({ id: item.VideoID });
        $.ajax('/Home/DeleteRow',
            {
                data: JSON,
                type: "POST",
                contentType: "application/json",
                success: function (data) {
                    self.Videos.remove(item);
                }
            });
    };

    //Load default data
    $.getJSON("/Home/JsonData", function (data) {
        ko.mapping.fromJS(data.Genres, {}, self.Genres);
        ko.mapping.fromJS(data.Authors, {}, self.Authors);
        var mappedVideo = $.map(data.Videos, function (item) { return new videoRow(item, data.Genres, data.Authors) });
        self.Videos(mappedVideo);
        $('#loader').hide();
        $('#table').show();
    }).error(function () { alert("Ошибка! Данные JSON не приняты!"); });

    self.datepicker = function () {
        $.datepicker.setDefaults($.datepicker.regional['ru']);
        $("input.date").datepicker();
    }
}

ko.applyBindings(new VideoViewModel());


